elytron/WFCORE-5120-client-default-ssl-context.adoc

```diff
@@ -0,0 +1,111 @@
+
+* The SSL context configured to match ALL rules[1] is the one that will be initialized and returned by this provider.
+
+[1] Elytron client can have multiple SSL contexts configured and `rules` are used to choose a single SSL context for the connection. Default SSL Context is the one that matches all rules.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
+== Implementation Plan
+
+New java security provider that provides SSLContext named "Default" will be implemented. This provider will load Elytron client configuration and will return initialized SSL context for users that request it with method `SSLContext.getDefault()`.
+
+It will be possible to pass Elytron client's configuration file path to the provider. If no path is specified, the provider will attempt to load the configuration from the current Elytron's authentication context obtained from classpath - this can be either `wildfly-config.xml` file on classpath or the file specified by system property `wildfly.config.url`.
+
+When this security provider is configured in JVM but there is no default SSL context configured in Elytron client, then this security provider will internally delegate the SSLContext instantiation to other providers. If no such other provider exists, NoSuchAlgorithmException exception will be thrown.
+
+== Test Plan
+
+Automatic smoke tests - placed in `wildfly-elytron` repository. Most important use cases whould be covered. This includes :
+ * A test case that passes configuration file path to the provider. Make sure it can load the configuration successfully and the provider returned from SSLContext.getDefault() is the Elytron security provider.
+ * A test case that generates a configuration from classpath. Make sure it can load the configuration successfully and the provider returned from SSLContext.getDefault() is the Elytron security provider.
+ * Test that when this security provider is configured in JVM but there is no default SSL context configured in Elytron client, then this security provider will internally delegate the SSLContext instantiation to other provider.
+Manual tests - statically configure `java.security` file to have this provider placed with the highest priority. Try to connect to the server that requires custom SSL context configured and make sure such request is successful when the configuration is on classpath and does not succeed when it is not.
+
+Once the default SSL context is set for the JVM, it cannot be overwritten. So to not mess with other tests in the testsuite, each automatic test for this provider will run in its own JVM that is terminated after the test finishes. This slows the testsuite. Because of this, only smoke tests should be put in the testsuite and other testing can be done manually.
```

Comment:
I can agree on keeping the WildFly TS free from some scenarios, e.g. complex configurations or ones that are expensive on the resources or that affect time.

But either the project (WIldFly) or the product (EAP) will need for this to be tested automatically.

Isn't there any other option for this? I didn't look at the code in detail, yet, but could forking the execution on a separate JVM process that surefire provides be an option? 
Or do you have any other alternatives in mind? 
Or we decide tat those are just manual tests, but it doesn't seem the case to me since this seems to be the only actual  integration test, here.

WDYT?

Reply:
@fabiobrz I added 3 test cases to elytron testsuite that test this automatically:
 *  1 that test this functionality when provider is taken from the SSL context from the current context
 *  1 where the provider has configuration file path passed as parameter
 *  1 where there is no specific default SSL contextconfigured in the current context and see if the returned SSL context is delegating to other providers. 
  
 IMHO these 3 automatic tests are enough to ensure the working of this provider?

Each of these 3 test is in the separate test class / file. This is why a new JVM is being created for each.